packages/aztec.js/src/proof/privateRange/index.js

```diff
@@ -0,0 +1,161 @@
+const proofUtils = require('../proofUtils');
+const verifier = require('./verifier');
+const Keccak = require('../../keccak');
+
+const privateRange = {};
+privateRange.verifier = verifier;
+
+const { groupReduction } = bn128;
+
+const { inputCoder, outputCoder } = require('../../abiEncoder');
+
+/**
+ * Construct blinding factors
+ *
+ * @method constructBlindingFactors
+ * @param {Object[]} notes AZTEC notes
+ * @param {Keccak} rollingHash structure containing the note data, to be hashe
+ * @returns {Object[]} blinding factors
+ */
+privateRange.constructBlindingFactors = (notes, rollingHash) => {
+    const bkArray = [];
+    let bk;
+    let B;
+    let x = new BN(0).toRed(groupReduction);
+
+    return notes.map((note, i) => {
+        const ba = bn128.randomGroupScalar();
+        bk = bn128.randomGroupScalar();
+
+        if (i === 0) {
+            B = note.gamma.mul(bk).add(bn128.h.mul(ba));
+        } else if (i === 1) {
+            x = rollingHash.keccak(groupReduction);
+            const xbk = bk.redMul(x);
+            const xba = ba.redMul(x);
+
+            B = note.gamma.mul(xbk).add(bn128.h.mul(xba));
+        } else if (i > 1) {
+            bk = bkArray[i - 2].redSub(bkArray[i - 1]);
+
+            x = rollingHash.keccak(groupReduction);
+            const xbk = bk.redMul(x);
+            const xba = ba.redMul(x);
+
+            B = note.gamma.mul(xbk).add(bn128.h.mul(xba));
+        }
+        bkArray.push(bk);
+
+        return {
+            bk,
+            ba,
+            B,
+        };
+    });
+};
+
+/**
+ * Construct AZTEC private range proof transcript
+ *
+ * @method constructProof
+ * @param {Object[]} notes AZTEC notes
+ * @param {string} sender the address calling the constructProof() function
+ * @returns {string[]} proofData - constructed cryptographic proof data
+ * @returns {string} challenge - crypographic challenge variable, part of the sigma protocol
+ */
+privateRange.constructProof = (notes, sender) => {
+    // rolling hash is used to combine multiple bilinear pairing comparisons into a single comparison
+    const rollingHash = new Keccak();
+
+    proofUtils.parseInputs(notes, sender);
+
+    notes.forEach((note) => {
+        rollingHash.append(note.gamma);
+        rollingHash.append(note.sigma);
+    });
+
+    const blindingFactors = privateRange.constructBlindingFactors(notes, rollingHash);
+
+    const challenge = proofUtils.computeChallenge(sender, notes, blindingFactors);
+    const proofData = blindingFactors.map((blindingFactor, i) => {
+        let kBar;
+
+        // Only set the first 2 values of kBar - the third is later inferred
+        // from a cryptographic relation. Set the third to 0
+        if (i <= 1) {
+            kBar = notes[i].k
+                .redMul(challenge)
+                .redAdd(blindingFactor.bk)
+                .fromRed();
+        } else {
+            kBar = padLeft(new BN(crypto.randomBytes(32), 16).umod(bn128.curve.n).toString(16), 64);
```

Comment:
We currently throw an error during proof construction if the scalars are zero. So, setting kBar to 0 here would cause proof construction to fail. 

I agree that this behaviour is desired though, will look into it